ciborium/src/ser/mod.rs

```diff
@@ -326,19 +380,109 @@ macro_rules! end {
+        fn end(self) -> Result<(), Self::Error> {
+            #[cfg(feature = "std")]
+            match self.serializer.canonicalization {
+                CanonicalizationScheme::None => {}
+                CanonicalizationScheme::Rfc8049 => {
+                    // keys get sorted in lexicographical byte order
+                    let keys = self.cache_keys;
+                    let values = self.cache_values;
+
+                    debug_assert_eq!(
+                        keys.len(), values.len(),
+                        "ciborium error: canonicalization failed, different number of keys and values?");
+
+                    let mut pairs: Vec<_> =
+                        keys.iter().zip(values.iter()).collect();
+
+                    pairs.sort();
+
+                    for (key, value) in pairs.iter() {
+                        self.serializer.encoder.write_all(&key)?;
+                        self.serializer.encoder.write_all(&value)?;
+                    }
+                }
+                CanonicalizationScheme::Rfc7049 => {
+                    // keys get sorted in length-first byte order
+                    let keys = self.cache_keys;
+                    let values = self.cache_values;
+
+                    debug_assert_eq!(
+                        keys.len(), values.len(),
+                        "ciborium error: canonicalization failed, different number of keys and values?");
+
+                    let mut pairs: Vec<_> =
+                        keys.iter()
+                            .map(|key| (key.len(), key))  // length-first ordering
+                            .zip(values.iter())
+                            .collect();
+
+                    pairs.sort();
+
+                    for ((_, key), value) in pairs.iter() {
+                        self.serializer.encoder.write_all(&key)?;
+                        self.serializer.encoder.write_all(&value)?;
+                    }
+                }
+            }
+
+            if self.indefinite {
+                self.serializer.encoder.push(Header::Break)?;
+            }
+
+            Ok(())
+        }
+    };
+}
+
+/// An internal struct for serializing collections.
+///
+/// Not to be used externally, only exposed as part of the [Serializer] type.
+#[doc(hidden)]
+pub struct CollectionSerializer<'a, W> {
+    serializer: &'a mut Serializer<W>,
+
+    /// Whether the collection is indefinite length. Cannot be used with canonical serialization.
+    indefinite: bool,
     tag: bool,
+
+    #[cfg(feature = "std")]
+    cache_keys: Vec<Vec<u8>>,
+    #[cfg(feature = "std")]
+    cache_values: Vec<Vec<u8>>,
+}
+
+impl<'a, W> CollectionSerializer<'a, W> {
+    pub fn new(serializer: &'a mut Serializer<W>, indefinite: bool, tag: bool) -> Self {
+        #[cfg(feature = "std")]
+        assert!(
+            !(serializer.canonicalization.is_sorting() && indefinite),
+            "ciborium error: canonical mode cannot be used with indefinite length collections");
```

Comment:
Noting a todo: for various reasons serde will give a zero length. The better option is to output the map/array headers during the end method as the length is known at that point. I can think of an easy way to do it in the macro, but don't want it to feel like too much of a hack.

Reply:
Should be resolved in the latest commit (force-pushed to fix the commit signing and message formatting, apologies!).